### PR TITLE
Wrap reflection with our own reflection class

### PR DESCRIPTION
**Pull request by BackEndTea**
This PR:

- [x] Wraps reflection with our own class
- [x] Covered by tests

This will make some things a bit easier: It allows us to make a switch to better reflection, removing our need to autoload the users code.

It also means we could possibly update the canMutate function to take this information as a param. 
e.g.

```php
public function canMutate(Node $node, InfectionReflectionClass $reflection): bool;
```

Once again ive chosen terrible names, and as im writing this i realize `InfectionReflectionClass` should probably be something like `ClassReflection`.

### CODE REVIEW

File: src/Reflection/CoreInfectionReflectionClass.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+use ReflectionException;
+
+/**
+ * @internal
+ */
+final class CoreInfectionReflectionClass implements InfectionReflectionClass
+{
+    private $reflectionClass;
+
+    public function __construct(ReflectionClass $reflectionClass)
```
Comment (theofidry) [inline src/Reflection/CoreInfectionReflectionClass.php]: Maybe you can add a static factory method to avoid to have to instantiate a `ReflectionClass` (as a consumer)?

File: src/Reflection/CoreInfectionReflectionClass.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        $this->reflectionClass = $reflectionClass;
+    }
+
+    public function hasParentOfVisibility(string $methodName, Visibility $visibility): bool
```
Comment (theofidry) [inline src/Reflection/CoreInfectionReflectionClass.php]: `hasParentMethodWithVisibility`?

File: src/Reflection/InfectionReflectionClass.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Reflection;
+
+/**
+ * @internal
+ */
+interface InfectionReflectionClass
```
Comment (theofidry) [inline src/Reflection/InfectionReflectionClass.php]: I'm really not convinced by having an interface here

File: src/Reflection/Visibility.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Reflection;
+
+/**
+ * @internal
+ */
+final class Visibility
```
Comment (theofidry) [inline src/Reflection/Visibility.php]: For me this class smells the "I want a real enum in PHP" even at the cost of working around the language too much. IMO just having a `public const ALL = [self::PUBLIC, self::PROTECTED]` instead of L46-L74 would be enough. You can add an `Assert::oneOf($visibility, Visibility::ALL)` statement if you wish.

File: tests/phpunit/Reflection/CoreInfectionReflectionClassTest.php [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    public function test_it_wraps_the_name(): void
+    {
+        $reflection = new CoreInfectionReflectionClass(new ReflectionClass($this));
+        $this->assertSame($reflection->getName(), get_class($this));
+    }
+
+    /**
+     * @dataProvider provideParentMethodCases
+     */
+    public function test_it_knows_if_a_function_is_inherited(CoreInfectionReflectionClass $reflection, string $method, Visibility $visibility, bool $hasParent): void
```
Comment (theofidry) [inline tests/phpunit/Reflection/CoreInfectionReflectionClassTest.php]: this like looks a bit long, maybe some line returns here for the parameters?

File: tests/phpunit/Reflection/CoreInfectionReflectionClassTest.php [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+use Generator;
+use Infection\Reflection\CoreInfectionReflectionClass;
+use Infection\Reflection\Visibility;
+use Infection\Visitor\CloneVisitor;
+use PHPUnit\Framework\TestCase;
+use ReflectionClass;
+
+final class CoreInfectionReflectionClassTest extends TestCase
+{
+    public function test_it_wraps_the_name(): void
```
Comment (theofidry) [inline tests/phpunit/Reflection/CoreInfectionReflectionClassTest.php]: `it_exposes_the_class_name`

File: src/Reflection/Visibility.php (hunk 4 above)
Comment (BackEndTea) [reply src/Reflection/Visibility.php]: Its probably a bit overkill, but now this class has an API that can't be worngly used.

Comment (BackEndTea) [reply src/Reflection/Visibility.php]: True, but now this class can't be used wrong, but i agree its a bit overkill

File: src/Reflection/InfectionReflectionClass.php (hunk 3 above)
Comment (BackEndTea) [reply src/Reflection/InfectionReflectionClass.php]: I have updated the PR, and there are now multiple implementation of this interface.

File: tests/phpunit/Mutator/AbstractMutatorTestCase.php [hunk 7]
```diff
@@ -145,7 +145,7 @@ protected function setUp(): void
         $mutants = [];
 
         foreach ($mutations as $mutation) {
-            $mutatorVisitor = new SimpleMutatorVisitor($mutation);
+            $mutatorVisitor = new MutatorVisitor($mutation);
```
Comment (BackEndTea) [inline tests/phpunit/Mutator/AbstractMutatorTestCase.php]: I ran into some issues with the `SimpleMutatorVisitor`.
Mainly that it calls 'canMutate' on every node again, which isn't something the 'normal' implementation does, so i've opted to replace it with the normal variant

File: src/Mutation/NodeTraverserFactory.php [hunk 8]
```diff
@@ -56,6 +56,10 @@ public function create(array $extraVisitors): PrioritizedVisitorsNodeTraverser
         $traverser = new PrioritizedVisitorsNodeTraverser(new NodeTraverser());
 
         $traverser->addPrioritizedVisitor(new NotMutableIgnoreVisitor(), 50);
+        $traverser->addPrioritizedVisitor(new NodeVisitor\NameResolver(null, [
```
Comment (theofidry) [inline src/Mutation/NodeTraverserFactory.php]: missing an import statement: we didn't use partial imports for the visitors in other places here

File: src/Mutator/IgnoreMutator.php [hunk 9]
```diff
@@ -82,10 +82,10 @@ public function canMutate(Node $node): bool
             return false;
         }
 
-        /** @var ReflectionClass|false $reflectionClass */
-        $reflectionClass = $node->getAttribute(ReflectionVisitor::REFLECTION_CLASS_KEY, false);
+        /** @var ClassReflection|null $reflectionClass */
+        $reflectionClass = $node->getAttribute(ReflectionVisitor::REFLECTION_CLASS_KEY);
 
-        if ($reflectionClass === false) {
+        if (!$reflectionClass instanceof ClassReflection) {
```
Comment (theofidry) [inline src/Mutator/IgnoreMutator.php]: `!($reflectionClass instanceof ClassReflection)`? I always find the `!` horribly confusing in this scenario

File: src/Reflection/AnonymousClassReflection.php [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+        return self::hasMethodRecursively($parent, $methodName, $visibility);
+    }
+
+    private static function hasMethod(
+        ReflectionClass $reflectionClass,
+        string $methodName,
+        Visibility $visibility
+    ): bool {
+        if ($reflectionClass->hasMethod($methodName)) {
```
Comment (theofidry) [inline src/Reflection/AnonymousClassReflection.php]: I would invert the expression to reduce the indent here

File: src/Reflection/CoreClassReflection.php [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+final class CoreClassReflection implements ClassReflection
+{
+    private $reflectionClass;
+
+    private function __construct(ReflectionClass $reflectionClass)
+    {
+        $this->reflectionClass = $reflectionClass;
+    }
+
+    public static function fromClassName(string $className): self
```
Comment (theofidry) [inline src/Reflection/CoreClassReflection.php]: I would say the static methods go before the non-static ones... But maybe it's something we should fix via the CS-Fixer rather

File: src/Reflection/Visibility.php (hunk 4 above)
Comment (theofidry) [reply src/Reflection/Visibility.php]: I think it's the kind of decision that fits better in a dedicated PR with an ADR entry tbh: it's not critical to the PR and should be relatively easy to enforce everywhere in one go

On a side note: if you wanted a close real enum, it would be preferable to implement the singleton pattern there to ease the comparison of two instances

File: tests/phpunit/Fixtures/SimpleMutation.php [hunk 12]
```diff
@@ -4,10 +4,12 @@
 
 namespace Infection\Tests\Fixtures;
 
+use Infection\MutatedNode;
+use Infection\Mutation\Mutation;
 use Infection\Mutator\Mutator;
 use PhpParser\Node;
 
-class SimpleMutation
+class SimpleMutation extends Mutation
```
Comment (theofidry) [inline tests/phpunit/Fixtures/SimpleMutation.php]: Might be worth looking in another PR if `SimpleMutation` is still necessary? I feel like a test factory for `Mutator` might be more appropriate than this class... if necessary at all

File: tests/phpunit/Visitor/ReflectionVisitorTest.php [hunk 13]
```diff
@@ -316,6 +338,10 @@ private function parseAndTraverse(string $code, ?NodeVisitor $nodeVisitor = null
         $this->traverse(
             $nodes,
             [
+                new NodeVisitor\NameResolver(null, [
```
Comment (theofidry) [inline tests/phpunit/Visitor/ReflectionVisitorTest.php]: missing import statement?

File: tests/phpunit/Visitor/ReflectionVisitorTest.php [hunk 14]
```diff
@@ -316,6 +339,10 @@ private function parseAndTraverse(string $code, ?NodeVisitor $nodeVisitor = null
         $this->traverse(
             $nodes,
             [
+                new NameResolver(null, [
```
Comment (theofidry) [inline tests/phpunit/Visitor/ReflectionVisitorTest.php]: Instead of having it repeated over, since it requires some special settings, what about having a static factory for it?

File: src/Reflection/Visibility.php (hunk 4 above)
Comment (theofidry) [reply src/Reflection/Visibility.php]: Still don't like it, but I'll leave it to the others to whether they are fine with it or would prefer the issue to be tackled in a dedicated PR

File: src/Visitor/ReflectionVisitor.php [hunk 15]
```diff
@@ -184,4 +182,26 @@ private function isFunctionLikeNode(Node $node): bool
...
+    private function getClassReflectionForNode(Node\Stmt\ClassLike $node): ClassReflection
+    {
+        /** @var Node\Name|null $fqn */
+        $fqn = $node->getAttribute(FullyQualifiedClassNameVisitor::FQN_KEY);
+
+        if ($fqn instanceof Node\Name) {
+            return CoreClassReflection::fromClassName($fqn->toString());
+        }
+
+        assert($node instanceof Node\Stmt\Class_);
```
Comment (theofidry) [inline src/Visitor/ReflectionVisitor.php]: I'm not sure to understand this assert neither how we get from `ClassLike` to `Class_` here... Is it based on the existence of the name?

Comment (BackEndTea) [reply src/Visitor/ReflectionVisitor.php]: If the classlike has no name then it is an anonymous class. This is the best way i could think of to make that clear to our tooling etc.
